ambry-frontend/src/test/java/com.github.ambry.frontend/AmbryBlobStorageServiceTest.java

```diff
@@ -1268,6 +1480,266 @@ private JSONObject createRequestHeaders(ByteRange range, GetOption getOption) th
+   * blobId string, for get/head/delete operations.
+   * @param blobId The blobId string to get/head/delete.
+   * @param expectedAccount The expected {@link Account} to verify its presence in {@link RestRequest}.
+   * @param expectedContainer The expected {@link Container} to verify its presence in {@link RestRequest}.
+   * @param expectedRestErrorCode The expected {@link RestServiceErrorCode} to verify.
+   * @throws Exception
+   */
+  private void verifyAccountAndContainerFromBlobId(String blobId, Account expectedAccount, Container expectedContainer,
+      RestServiceErrorCode expectedRestErrorCode) throws Exception {
+    RestRequest restRequest;
+    for (RestMethod restMethod : Lists.newArrayList(RestMethod.GET, RestMethod.HEAD, RestMethod.DELETE)) {
+      restRequest = createRestRequest(restMethod, blobId, null, null);
+      MockRestResponseChannel restResponseChannel = new MockRestResponseChannel();
+      try {
+        doOperation(restRequest, restResponseChannel);
+        if (expectedRestErrorCode != null) {
+          if (expectedRestErrorCode != null) {
+            fail("Should have thrown");
+          }
+        }
+      } catch (RestServiceException e) {
+        assertEquals("Wrong RestServiceErrorCode", expectedRestErrorCode, e.getErrorCode());
+      }
+      assertEquals("Wrong account object in RestRequest's args", expectedAccount,
+          restRequest.getArgs().get(RestUtils.InternalKeys.TARGET_ACCOUNT_KEY));
+      assertEquals("Wrong container object in RestRequest's args", expectedContainer,
+          restRequest.getArgs().get(RestUtils.InternalKeys.TARGET_CONTAINER_KEY));
+      assertEquals("Wrong account object in RestRequest's args", expectedAccount,
+          restRequest.getArgs().get(RestUtils.InternalKeys.TARGET_ACCOUNT_KEY));
+      assertEquals("Wrong container object in RestRequest's args", expectedContainer,
+          restRequest.getArgs().get(RestUtils.InternalKeys.TARGET_CONTAINER_KEY));
+    }
+  }
+
+  /**
+   * Puts a blob and verifies the injected {@link Account} and {@link Container} into the {@link RestRequest}.
+   * @param accountName The accountName to send as the header of the request.
+   * @param containerName The containerName to send as the header of the request.
+   * @param serviceId The serviceId to send as the header of the request.
+   * @param isPrivate The isPrivate flag for the blob.
+   * @param expectedAccount The expected {@link Account} that would be injected into the {@link RestRequest}.
+   * @param expectedContainer The expected {@link Container} that would be injected into the {@link RestRequest}.
+   * @param expectedRestErrorCode The expected {@link RestServiceErrorCode} after the put operation.
+   * @return The blobId string if the put operation is successful, {@link null} otherwise.
+   * @throws Exception
+   */
+  private String putBlobAndVerifyWithAccountAndContainer(String accountName, String containerName, String serviceId,
```

Comment:
I think this is a valid point. I was trying to reuse `postBlobAndVerify`, but it seems I will still go through similar amount of code changes because of the error code check logic. So I would prefer to keep this method for now, which stay logically isolated (This method does not check other fields).